src/core/reaction_methods/ReactionAlgorithm.cpp

```diff
@@ -0,0 +1,651 @@
+#include <limits>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace ReactionMethods {
+
+/**
+ * Performs a randomly selected reaction in the reaction ensemble
+ */
+int ReactionAlgorithm::do_reaction(int reaction_steps) {
+  for (int i = 0; i < reaction_steps; i++) {
+    int reaction_id = i_random(static_cast<int>(reactions.size()));
+    generic_oneway_reaction(reaction_id);
+  }
+  return 0;
+}
+
+/**
+ * Adds a reaction to the reaction system
+ */
+void ReactionAlgorithm::add_reaction(
+    double gamma, const std::vector<int> &reactant_types,
+    const std::vector<int> &reactant_coefficients,
+    const std::vector<int> &product_types,
+    const std::vector<int> &product_coefficients) {
+  SingleReaction new_reaction(gamma, reactant_types, reactant_coefficients,
+                              product_types, product_coefficients);
+
+  // make ESPResSo count the particle numbers which take part in the reactions
+  for (int reactant_type : new_reaction.reactant_types)
+    init_type_map(reactant_type);
+  for (int product_type : new_reaction.product_types)
+    init_type_map(product_type);
+
+  init_type_map(non_interacting_type);
+
+  reactions.push_back(new_reaction);
+}
+
+/**
+ * Checks whether all necessary variables for the reaction ensemble have been
+ * set.
+ */
+void ReactionAlgorithm::check_reaction_ensemble() const {
+  if (reactions.empty()) {
+    throw std::runtime_error("Reaction system not initialized");
+  }
+
+  if (temperature < 0) {
+    throw std::runtime_error("Temperatures cannot be negative. Please provide "
+                             "a temperature (in k_B T) to the simulation. "
+                             "Normally it should be 1.0. This will be used "
+                             "directly to calculate beta:=1/(k_B T) which "
+                             "occurs in the exp(-beta*E)\n");
+  }
+#ifdef ELECTROSTATICS
+  // check for the existence of default charges for all types that take part in
+  // the reactions
+
+  for (const auto &current_reaction : reactions) {
+    // check for reactants
+    for (int reactant_type : current_reaction.reactant_types) {
+      auto it = charges_of_types.find(reactant_type);
+      if (it == charges_of_types.end()) {
+        std::string message = std::string("Forgot to assign charge to type ") +
+                              std::to_string(reactant_type);
+        throw std::runtime_error(message);
+      }
+    }
+    // check for products
+    for (int product_type : current_reaction.product_types) {
+      auto it = charges_of_types.find(product_type);
+      if (it == charges_of_types.end()) {
+        std::string message = std::string("Forgot to assign charge to type ") +
+                              std::to_string(product_type);
+        throw std::runtime_error(message);
+      }
+    }
+  }
+#endif
+}
+
+/**
+ * Automatically sets the volume which is used by the reaction ensemble to the
+ * volume of a cuboid box
```

Comment:
If it was not initialized with some other value

Reply:
done